library/math/dual.lisp

```diff
@@ -0,0 +1,177 @@
+;;;; dual.lisp
+
+
+(coalton-library/utils:defstdlib-package #:coalton-library/math/dual
+  (:use
+   #:coalton
+   #:coalton-library/builtin
+   #:coalton-library/classes
+   #:coalton-library/functions
+   #:coalton-library/math/arith
+   #:coalton-library/math/elementary
+   #:coalton-library/math/integral
+   #:coalton-library/hash)
+  (:export
+   #:Dual
+   #:primal-part
+   #:dual-part)
+  (:documentation "
+dual numbers are a hypercomplex number system [1]. A dual number has the form
+a + bε where a and b are real numbers and ε is a symbol that satisfies ε^2=0
+and ε!=0. One application of dual numbers is automatic differentiation; an example
+taken from [2] is as follows:
+
+    consider you have the given expression f(x) = 3x+2 and you want to calculate
+    f(4) and f'(4).
+
+    You then convert 4 into a dual number which gives:
+
+        4+1ε
+
+    then the computation is as follows:
+
+        (4*1ε) * (3 + 0ε) =
+        12 * 0ε + 3ε + 0ε^2 =
+        12 + 3ε
+
+    lastly, you need to add the constant 2 of the expression
+
+        (12 + 3ε) + (2 + 0ε) =
+        14 + 3ε
+
+   in this result, the primal 14 is the value of f(4) and the dual is the value of
+   of f'(4).
+
+   Haskell has an automatic differentiation library and you can find it here
+   https://hackage.haskell.org/package/ad.
+
+   limitations:
+
+     some limitations is that `Ord` and `Eq` and `Hash` only work on primals.
+     but the advantages of this approach is that code will be auto differentiated
+     for free.
+
+   references:
+
+   [1] https://en.wikipedia.org/wiki/Dual_number
+   [2] https://blog.demofox.org/2014/12/30/dual-numbers-automatic-differentiation/"))
+
+(in-package #:coalton-library/math/dual)
+
+(named-readtables:in-readtable coalton:coalton)
+
+#+coalton-release
+(cl:declaim #.coalton-impl/settings:*coalton-optimize-library*)
+
+(coalton-toplevel
+
+  (define-type (Dual :t)
+    "Representation of a Dual number in the form (a + bε) where a and b are  real numbers and ε satisfie ε**2 = 0 and ε != 0.  Note: `Eq`, and `Ord` and `Hash` only work on the primal component."
+    (Dual :t :t))
+   
+  (declare primal-part (Dual :t -> :t))
+  (define (primal-part (Dual p _))
+    p)
+
+  (declare dual-part (Dual :t -> :t))
+  (define (dual-part (Dual _ d))
+    d)
+
+  (define (sq x)
+    (* x x))
+   
+  (define-instance (Eq :t => Eq (Dual :t))
+    "Note: Eq only works on the primal components."
+    (define (== (Dual a _) (Dual p _))
+      (== a p)))
+  
+  (define-instance (Num :t => Num (Dual :t)) 
+    (define (+ (Dual p1 d1) (Dual p2 d2))
+      (Dual (+ p1 p2) (+ d1 d2)))
+     
+    (define (- (Dual p1 d1) (Dual p2 d2))
+      (Dual (- p1 p2) (- d1 d2)))
+
+    (define (* (Dual p1 d1) (Dual p2 d2))
+      (Dual (* p1 p2)
+            (+ (* p1  d2) (* d1 p2))))
+     
+    (define (fromInt z)
+      (Dual (fromInt z) 0)))
+   
+  (define-instance (Reciprocable :t => Reciprocable (Dual :t))
```

Comment:
In many functions below, I think it would be clearer to have functions like

```
(defun (d-recip x)
  (negate (recip (sq x))))

(defun (d-sin x)
  (cos x))

(defun (d-cos x)
  (negate (d-sin x)))
```

and so on. It would make implementing the instances a lot more clear and easier to review.

It's not required for this PR but I'd recommend it for future work.